posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc

```diff
@@ -0,0 +1,52 @@
+= "Calling JDK Tools Programmatically on Java 9"
+Gunnar Morling
+:awestruct-tags: [ "Discussions" ]
+:awestruct-layout: blog-post
+---
+
+Without doubt, the module system (JPMS) is the most prominent feature of Java 9. But there are many other useful additions to the JDK, which have not been discussed as widely yet.
+
+One of them is the new `ToolProvider` SPI, which defines a uniform way to invoke all the tools coming with the JDK (e.g. `jar`, `jlink` etc.) in a programmatic way.
+E.g. imagine you'd like to create a JAR archive from within your Java application.
+Before Java 9, you had two options to do so:
+
+* Start a separate process to run the `jar` binary
+* Find out what classes are used by the `jar` command internally and invoke them directly
+
+Both options aren't ideal.
+The former incurs some overhead for forking an OS level process and it requires a bit of coding to locate the binary to execute in the Java home directory and to correctly capture any output of the forked process.
+The second option doesn't come with the disadvantages of forking a new process, but it requires knowledge about the implementation of the tool to execute.
+So for instance in the case of the `jar` command, you'd have to know that it uses `java.util.jar.JarOutputStream` internally in order to call it directly.
```

Comment:
"it uses" ? Might be confusing as it sounds like `jar` *depends* on JarOutputStream. I take it you mean that JarOutputStream is its entry point?